### PR TITLE
Stop being so bail-y in candidate assembly

### PR DESCRIPTION
**Pull request by compiler-errors**
A conceptual follow-up to #132084. We gotta stop bailing so much when there are errors; it's both unnecessary, leads to weird knock-on errors, and it's messing up the vibes lol

### CODE REVIEW

**Comment by compiler-errors** [comment]
Whoops, I wanted an lcnr vibe check

r? lcnr

File: compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs [hunk 1]
```diff
@@ -91,14 +91,6 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             } else if tcx.is_lang_item(def_id, LangItem::Sized) {
                 // Sized is never implementable by end-users, it is
                 // always automatically computed.
-
-                // FIXME: Consider moving this check to the top level as it
-                // may also be useful for predicates other than `Sized`
-                // Error type cannot possibly implement `Sized` (fixes #123154)
-                if let Err(e) = obligation.predicate.skip_binder().self_ty().error_reported() {
-                    return Err(SelectionError::Overflow(e.into()));
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs:99]: :((((((((((( what the heck were we doing here!!

File: tests/ui/traits/object/dyn-no-impl-subst.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#![feature(dyn_compatible_for_dispatch)]
+
```
Comment (lcnr) [inline tests/ui/traits/object/dyn-no-impl-subst.rs:2]: regression test for X

**Comment by compiler-errors** [comment]
:think: i guess it's failing on `--target=i686-unknown-linux-gnu`.

**Pull request merged**